lib/index.js

```diff
@@ -65,28 +65,76 @@ exports.runOperation = async ({operation}) => {
     }
   }
 
-  if(isGenerateKeyOp) {
-    // TODO: handle cornercase where record insert worked previously but the
-    // plugin did not generate the key... currently, we insert the record
-    // first in case we crash before using the plugin but perhaps this should
-    // be inverted once this TODO is handled
+  // FIXME: remove comments
+  /*
+  client generates key ->
+
+  1. key pending key record is created in bedrock-kms
+  <failure> ssm request to generate key is never made
+  result, pending record in kms and no key record in ssm
+  state:
+    kms -> pending key
+    ssm -> no key
+
+  2. key pending key record is created in bedrock-kms
+  ssm request to generate key succeeds
+  <failure> kms does not record successful key generation (remove pending flag)
+  state:
+    kms -> pending key
+    ssm -> proper key
+
+  3. key pending key record is created in bedrock-kms
+  ssm request to generate key fails
+  key record is removed from bedrock-kms
+  state:
+    kms -> no key
+    ssm -> no key
+
+  4. key pending key record is created in bedrock-kms
+  ssm request to generate key fails
+  <failure> key record is not removed from bedrock-kms
+  state:
+    kms -> pending key
+    ssm -> no key
+    * sames as #1
+  */
+
+  let recoveryMode = false;
+  if(record && record.meta.pending) {
+    recoveryMode = true;
+  }
+
+  if(recoveryMode && !isGenerateKeyOp) {
+    // the caller was never told that the key generation succeeded
+    // operations on pending keys are not allowed
+    throw new BedrockError(
+      'Key not found.',
+      'NotFoundError',
+      {key: keyId, httpStatusCode: 404, public: true});
+  }
 
-    // check for duplicate key
-    if(record) {
+  // if there is an existing key record, controller must match in all cases
+  if(record) {
+    if(record.key.controller !== controller) {
+      throw new BedrockError(
+        'Permission denied.',
+        // FIXME: this was a NotFoundError in the case of non-generation
+        // operations, was that important, possible better security?
```

Comment:
Yeah, this was not found for that purpose ... not sure if other mechanisms could just get around that though. If not, we should put it back, but if it wasn't helpful, we can leave it as permission denied.

Reply:
Also, if it we change it to permission denied, the status code should be 400 not 404.

Reply:
Seems like a timing analysis would easily reveal this deception.  I think we go with permission denied, which is the simplest implementation unless there's an objection.